guide/oauth2/README.md

```diff
@@ -0,0 +1,200 @@
+
+```js
+function generateRandomString() {
+	const rand = Math.floor(Math.random() * 10);
+	let randStr = '';
+	for (let i = 0; i < 20 + rand; i++) {
+		randStr += String.fromCharCode(33 + Math.floor(Math.random() * 94));
+	}
+	return randStr;
+}
+
+// ...
+
+// generate and store the string
+const randStr = generateRandomString();
+localStorage.setItem('stateParameter', randStr);
+oauthURL += `&state=${btoa(randStr)}`;
+document.getElementById('login').href = oauthURL;
+```
+
+When you visit a url with a state parameter appended to it and then click `Authorize`, you'll notice that after being redirected, the url will also have the state parameter appended to it, which you should then check against what was stored.
+
+```js
+const stateParameter = localStorage.getItem('stateParameter');
+if (btoa(stateParameter) !== urlStateParameter) {
+	// you may have been clickjacked!
+}
+```
+
+<p class="tip">Don't forgo security for a tiny bit of convenience!</p>
+
+### OAuth2 flows
+
+What we did in our quick example was go through the `implicit grant` flow, which passed the access token straight to the user's browser. This is great and simple, but you don't get to refresh the token without the user and it is less secure than going through the `authorization code grant`. This involves recieving an access code, which is then exchanged by your server for an access token. Notice that this way, the access token never actually reaches the user throughout the process.
+
+#### Authorization code grant
+
+Unlike the quick example, we need an OAuth2 url where the `response_type` is `code`. Once you've obtained it, try visiting the link and authorizing your application. You should notice that instead of a hash, the redirect url now has a single query parameter appended to it like `?code=ACCESS_CODE`. Let's modify our `index.js` file to pull the parameter out of the url if it exists.
+
+```js
+const [url, queryParams] = req.url.split('?');
+if (queryParams) {
+	const accessCode = queryParams.match(/code=(.+?)(?:&|$)/);
+	if (accessCode) {
+		console.log(`The access code is: ${accessCode[1]}`);
+	}
+}
+if (url === '/') {
+	responseCode = 200;
+	content = fs.readFileSync('./index.html');
+}
+```
+
+Now, that you have to exchange this code with Discord for an access token. To do this, you need your `client_id` and `client_secret`. If you've forgotten them, head over to [your applications](https://discordapp.com/developers/applications) and get them. We'll be using `node-fetch` along with `form-data` to make our request to Discord; `node-fetch` should be bundled with the latest versions of Discord.js, and you can install `form-data` with `npm i form-data`.
```

Comment:
Regarding: "`node-fetch` should be bundled with the latest versions of Discord.js"
This isn't exactly true because the guide bases its content based on the stable branch, so the user should still be told to install `node-fetch`.

This'll be a non-issue once #121 becomes a thing. Soon:tm: